src/Algebra/Module/Construct/Nagata.agda

```diff
@@ -0,0 +1,170 @@
+------------------------------------------------------------------------
+-- The Agda standard library
+--
+-- The non-commutative analogue of Nagata's construction
+-- of the "idealization of a module", defined here
+-- on R-R-*bi*modules M over a ring R.
+--
+-- Elements of R ⋉ M are pairs |R| × |M|
+------------------------------------------------------------------------
+
+{-# OPTIONS --cubical-compatible --safe #-}
+
+module Algebra.Module.Construct.Nagata where
+
+open import Algebra.Bundles
+open import Algebra.Core
+import Algebra.Consequences.Setoid as Consequences
+import Algebra.Definitions as Definitions
+open import Algebra.Module.Bundles
+open import Algebra.Module.Core
+open import Algebra.Module.Definitions
+import Algebra.Module.Construct.DirectProduct as DirectProduct
+import Algebra.Module.Construct.TensorUnit as TensorUnit
+open import Algebra.Structures
+open import Data.Product using (_×_; _,_; proj₁; proj₂)
+open import Data.Product.Relation.Binary.Pointwise.NonDependent
+open import Level using (Level; _⊔_)
+open import Relation.Binary using (Rel; Setoid; IsEquivalence)
+import Relation.Binary.Reasoning.Setoid as ≈-Reasoning
+
+private
+  variable
+    m ℓm r ℓr : Level
+
+------------------------------------------------------------------------
+-- Definitions
+
+module Nagata (ring : Ring r ℓr) (bimodule : Bimodule ring ring m ℓm) where
```

Comment:
We can lift these to top-level module parameters (see comment about the infix notation below for what to do with it).

Reply:
So this is an interesting one.
As it stands, the construction appears indeed to only take top-level parameters, so could be parametrised at the top-level as such. But I imagined a time, after later refactoring, when all the (semi-)etc. variants of this construction might all be brought under one roof: mind you, my instincts are for smaller modules, so I recoil in horror somewhat at the thought of it (plus being uncertain exactly what axiomatisation yields what; this is already a modest generalisation of what I could find in the literature, which typically focuses on case when R is commutative).

Will consider refactoring when I've done the lower-hanging stuff in yours, and @JacquesCarette 's reviews.

Reply:
More or less just this one to do, but still wondering whether it's a good idea?